plugins/state_history_plugin/include/eosio/state_history_plugin/session.hpp

```diff
@@ -16,691 +19,327 @@ extern const char* const state_history_plugin_abi;
+      catch(...) {
+         if(has_logged_exception.test_and_set() == false)
+            fc_ilog(logger, "state history connection from ${a} failed", ("a", remote_endpoint_string));
       }
 
-      // not just an optimization, on accepted_block signal may not be able to find block_num in forkdb as it has not been validated
-      // until after the accepted_block signal
-      std::optional<chain::block_id_type> block_id =
-          (block && block->block_num() == to_send_block_num) ? id : plugin.get_block_id(to_send_block_num);
-
-      if (block_id && position_it && (*position_it)->block_num == to_send_block_num) {
-         // This branch happens when the head block of nodeos is behind the head block of connecting client.
-         // In addition, the client told us the corresponding block id for block_num we are going to send.
-         // We can send the block when the block_id is different.
-         auto& itr = *position_it;
-         auto block_id_seen_by_client = itr->block_id;
-         ++itr;
-         if (itr == request.have_positions.end())
-            position_it.reset();
-
-         if(block_id_seen_by_client == *block_id) {
-            ++to_send_block_num;
-            session_mgr.pop_entry(false);
-            return;
+      drop_exceptions([this](){ stream.next_layer().close(); });
+      drop_exceptions([this](){ awake_if_idle();             });
+   }
+
+   boost::asio::awaitable<void> read_loop() {
+      co_await readwrite_coro_exception_wrapper([this]() -> boost::asio::awaitable<void> {
+         wake_timer.expires_at(std::chrono::steady_clock::time_point::max());
+
+         if constexpr(std::is_same_v<SocketType, boost::asio::ip::tcp::socket>)
+            stream.next_layer().set_option(boost::asio::ip::tcp::no_delay(true));
+         stream.next_layer().set_option(boost::asio::socket_base::send_buffer_size(1024*1024));
+         stream.write_buffer_bytes(512*1024);
+         stream.set_option(boost::beast::websocket::stream_base::decorator([](boost::beast::websocket::response_type& res) {
+            res.set(boost::beast::http::field::server, "state_history/" + app().version_string());
+         }));
+
+         co_await stream.async_accept();
+         co_await stream.async_write(boost::asio::const_buffer(state_history_plugin_abi, strlen(state_history_plugin_abi)));
+         stream.binary(true);
+         boost::asio::co_spawn(strand, write_loop(), [&](std::exception_ptr e) {check_coros_done(e);});
+
+         while(true) {
+            boost::beast::flat_buffer b;
+            co_await stream.async_read(b);
+            const state_request req = fc::raw::unpack<std::remove_const_t<decltype(req)>>(static_cast<const char*>(b.cdata().data()), b.size());
+
+            //TODO: how can set main thread priority on this?
+            auto& self = *this; //gcc10 ICE workaround wrt capturing 'this' in a coro
+            co_await boost::asio::co_spawn(app().get_io_service(), [&]() -> boost::asio::awaitable<void> {
```

Comment:
as a minor nitpick, using an anonymous lambda here makes it harder to understand its meaning at a glance. 

even something simple like this may add readability:
```
auto add_requests_to_main_thread = [&]() -> boost::asio::awaitable<void> {
   ...
}
co_await boost::asio::co_spawn(app().get_io_service(), add_requests_to_main_thread, boost::asio::use_awaitable);
```

I'd also accept a doc-style function header inside the body.

Reply:
My preference of those two options is adding a comment, which I did in edb242d. I think the real strength of the coroutines is keeping code presented in the same order of execution and the inline anonymous lambda does better at that than "ping-ponging" back to a named lambda, or calling another function entirely (which starts becoming more callback-ish)

That said, I think there may be some sloppiness here around the capture-all aspect of the lambda; specifically how the main thread coro accesses variables in the outer coro to "fill them out" (as a way of returning values). It seems like maybe the main thread coro should return these values instead, for example instead of
```c++
std::deque<bool>             status_requests;
std::optional<block_package> block_to_send;
co_await boost::asio::co_spawn(app().get_io_service(), [&]() -> boost::asio::awaitable<void> {
   status_requests = std::move(self.queued_status_requests);
   // ...
   co_return; //void
```
it actually returns the values. for example,
```c++
auto [status_requests, block_to_send] = co_await boost::asio::co_spawn(app().get_io_service(), [&]() -> boost::asio::awaitable<std::pair<std::deque<bool>, std::optional<block_package>>> {
   std::deque<bool> status_requests_ret = std::move(self.queued_status_requests);
   // ...
   co_return std::make_pair(status_requests_ret, block_to_send_ret)
```
But, I wasn't really able to come up with an attractive way to do this in the little time I spent looking at it, and I wasn't entirely confident about move semantics and/or copy elision mixed together with coroutines either, without more investigation.